### PR TITLE
ISSUE 371 - Mailet to collect contacts - Optimize reactor & Assign sender is owner of contact

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala [hunk 1]
```diff
@@ -69,21 +72,20 @@ class ContactsCollection @Inject()(eventBus: EventBus) extends GenericMailet {
...
+        username = Username.fromMailAddress(sender),
         contact = contact),
         NO_REGISTRATION_KEYS)))
       .collectSeq()
-      .block()
+      .`then`()
 
-  private def appendAttributeToMail(mail: Mail, contacts: Seq[ContactFields]): Unit =
-    mail.getMaybeSender.asOptional().toScala
+  private def appendAttributeToMail(sender: MailAddress, mail: Mail, contacts: Seq[ContactFields]) =
```
Comment (Arsnael) [inline tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:85]: No method signature anymore?

**Review by Arsnael — APPROVED** [review]
For me it would make sense. Let's see what the others think :)

File: tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala [hunk 2]
```diff
@@ -55,9 +55,12 @@ class ContactsCollection @Inject()(eventBus: EventBus) extends GenericMailet {
 
   override def service(mail: Mail): Unit =
     if (CollectionUtils.isNotEmpty(mail.getRecipients)) {
-      val contacts: Seq[ContactFields] = extractRecipientsContacts(mail)
-      dispatchEvents(contacts)
-      appendAttributeToMail(mail, contacts)
+     SMono.justOrEmpty(mail.getMaybeSender.asOptional().toScala)
+        .map(sender => extractRecipientsContacts(mail) -> sender)
+        .flatMap {
+          case (contacts, sender) => SFlux.zip(appendAttributeToMail(sender, mail, contacts), dispatchEvents(sender, contacts))
```
Comment (quantranhong1999) [inline tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:61]: Should we check if the Sender is a user of James before dispatching the Event to EventBus?

I think we would not want to index stuff for outside sender like `elonmusk@tesla.com`.

Comment (chibenwa) [reply tmail-backend/mailets/src/main/scala/com/linagora/tmail/mailets/ContactsCollection.scala:61]: No not really generally this is coupled with SenderIs matcher

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**